Zend/zend_weakrefs.c

```diff
@@ -0,0 +1,217 @@
+}
+
+void zend_weakrefs_notify(zend_object *object) {
+	zend_hash_index_del(&EG(weakrefs), (zend_ulong) object);
+}
+
+void zend_weakrefs_shutdown() {
+	zend_hash_destroy(&EG(weakrefs));
+}
+
+static zend_object* zend_weakref_clone(zval *zv) {
+	Z_ADDREF_P(zv);
+
+	return Z_OBJ_P(zv);
+}
+
+static zend_object* zend_weakref_new(zend_class_entry *ce) {
+	zend_weakref *wr = zend_object_alloc(sizeof(zend_weakref), zend_ce_weakref);
+
+	zend_object_std_init(&wr->std, zend_ce_weakref);
+
+	wr->std.handlers = &zend_weakref_handlers;
+
+	return &wr->std;
+}
+
+static zend_always_inline zend_bool zend_weakref_find(zval *referent, zval *return_value) {
+	zend_weakref *wr = zend_hash_index_find_ptr(&EG(weakrefs), (zend_ulong) Z_OBJ_P(referent));
+
+	if (!wr) {
+		return 0;
+	}
+
+	GC_ADDREF(&wr->std);
+	ZVAL_OBJ(return_value, &wr->std);
+
+	return 1;
+}
+
+static zend_always_inline void zend_weakref_create(zval *referent, zval *return_value) {
+	zend_weakref *wr;
+
+	object_init_ex(return_value, zend_ce_weakref);
+
+	wr = zend_weakref_fetch(return_value);
+
+	ZVAL_COPY_VALUE(&wr->referent, referent);
+
+	zend_hash_index_add_ptr(&EG(weakrefs), (zend_ulong) Z_OBJ_P(referent), wr);
+
+	GC_ADD_FLAGS(Z_OBJ_P(referent), IS_OBJ_WEAKLY_REFERENCED);
+}
+
+static zend_always_inline void zend_weakref_get(zval *weakref, zval *return_value) {
+	zend_weakref *wr = zend_weakref_fetch(weakref);
+
+	ZVAL_COPY(return_value, &wr->referent);
+}
+
+static void zend_weakref_free(zend_object *zo) {
+	zend_weakref *wr = zend_weakref_from(zo);
+
+	if (Z_TYPE(wr->referent) == IS_OBJECT) {
+		zend_hash_index_del(
+			&EG(weakrefs), (zend_ulong) Z_OBJ(wr->referent));
+	}
+
+	zend_object_std_dtor(&wr->std);
+}
+
+#define zend_weakref_unsupported(thing) \
+	zend_throw_error(NULL, "WeakReference objects do not support " thing);
+
+static zval* zend_weakref_no_write(zval *object, zval *member, zval *value, void **rtc) {
+	zend_weakref_unsupported("properties");
+
+	return &EG(uninitialized_zval);
+}
+
+static zval* zend_weakref_no_read(zval *object, zval *member, int type, void **rtc, zval *rv) {
+	if (!EG(exception)) {
+		zend_weakref_unsupported("properties");
+	}
+
+	return &EG(uninitialized_zval);
+}
+
+static zval *zend_weakref_no_read_ptr(zval *object, zval *member, int type, void **rtc) {
+	zend_weakref_unsupported("property references");
+	return NULL;
+}
+
+static int zend_weakref_no_isset(zval *object, zval *member, int hse, void **rtc) {
+	if (hse != 2) {
+		zend_weakref_unsupported("properties");
+	}
+	return 0;
+}
+
+static void zend_weakref_no_unset(zval *object, zval *member, void **rtc) {
+	zend_weakref_unsupported("properties");
+}
+
+ZEND_BEGIN_ARG_WITH_RETURN_OBJ_INFO_EX(zend_weakref_of_arginfo, 0, 1, WeakReference, 0)
+	ZEND_ARG_TYPE_INFO(0, referent, IS_OBJECT, 0)
+ZEND_END_ARG_INFO()
+
+ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(zend_weakref_get_arginfo, 0, 0, IS_OBJECT, 1)
+ZEND_END_ARG_INFO()
+
+ZEND_COLD ZEND_METHOD(WeakRef, __construct)
```

Comment:
Rename `WeakRef` to `WeakReference` in the ZEND_METHOD declarations to make it more obvious what the class name is?